src/ecdsa/keygen.rs

```diff
@@ -695,6 +699,31 @@ impl State<KeyGeneratorTraits> for Phase2 {
         }]))
     }
 }
+/// Computes the sum of points on the curve and validates every point
+/// Uses PublicKey type from underlying secp256k1 library to work around limited API in curv crate
+fn try_computing_public_key(
```

Comment:
How about a more idiomatic version? Btw this version also handles the case when combination is unsuccessful.
```rust
fn try_computing_public_key2(
    pubkey_map: &HashMap<PartyIndex, GE>,
) -> Result<curv::PK, Vec<KeygenError>> {
    let acc: Result<Option<curv::PK>, Vec<KeygenError>> = Ok(None);

    let fold = pubkey_map
        .iter()
        .fold(acc, |mut acc, point| {
            if !is_valid_public_key(point.1.get_element()) {
                // Invalid pub key, record an error, discard any OK result so far.
                // Any further reconstruction of PK is meaningless,
                // combinations may succeed but it will not be an expected PK.
                let err = KeygenError::InvalidPublicKey {
                    point: format!("{:?} {:?}", point.1.get_element(), pubkey_map),
                    party: *point.0,
                };

                acc = match acc {
                    Ok(_) => Err(vec![err]),
                    Err(mut errs) => {
                        errs.push(err);
                        Err(errs)
                    }
                }
            }

            // No errors encountered so far. PK reconstruction may succeed
            let combination = acc.map(|partial_pk|
                partial_pk.map_or_else(
                    || Ok(point.1.get_element()),
                    |pk| pk.combine(&point.1.get_element())
                )
            );

            match combination {
                Ok(Ok(sum)) => Ok(Some(sum)),
                Ok(Err(err)) => {
                    // No errors were encountered until the combination attempt.
                    Err(vec![
                        // Insert appropriate error, next line is a placeholder.
                        KeygenError::GeneralError("cant reconstruct public key".to_string())
                    ])
                },
                // Propagate errors.
                Err(errs) => Err(errs)
            }
        });

    match fold {
        Ok(Some(pk)) => Ok(pk),
        Ok(None) => Err(vec![
            // Insert appropriate error, next line is a placeholder.
            KeygenError::GeneralError("public key map is empty".to_string())
        ]),
        Err(errs) => Err(errs)
    }
}
```

Reply:
This looks absolutely fantastic, although original code is ~3 times shorter and it does not use cumbersome Option<Result> and Result<Option> types. 
And why do you think that it does not handle  combination errors?

Reply:
Consider this
```rust
if let Ok(sum) = v.combine(&point.1.get_element()) {
                        return (Some(sum), evec);
                    }
```
what happens if `v.combine(&point.1.get_element())` fails?

Reply:
Do you agree that the code falls through down to evec.push() in case if let Ok() cond is not satisfied,  or am I missing something in the question? Any evec content will trigger machine exit , line 834 so that the snippet you mention will not be executed.

Reply:
Indeed, you're right.
Although, the erroneous value of `v.combine` gets replaced by `KeygenError::InvalidPublicKey`. Let's assume this loss is not critical.
Still, I find a triple nested conditionals not nice to the very least. Consider this version
```rust
if is_valid_public_key(point.1.get_element()) {
                let acc = acc.map_or_else(
                    || Ok(point.1.get_element()),
                    |v| v.combine(&point.1.get_element())
                );

                if let Ok(sum) = acc {
                    return (Some(sum), evec)
                }
            }

            evec.push(KeygenError::InvalidPublicKey {
                point: format!("{:?} {:?}", point.1.get_element(), pubkey_map),
                party: *point.0,
            });
            
            (acc, evec)
```
IMO: easier to read and can be made shorter by inlining deconstruction of `acc.map_or_else`.